app/assets/javascripts/discourse/app/lib/highlight-syntax.js

```diff
@@ -23,25 +24,85 @@ export default function highlightSyntax(elem, siteSettings, session) {
 
   const path = session.highlightJsPath;
 
-  if (!path) {
-    return;
-  }
+  const hljs = await ensureHighlightJs(path);
 
-  return loadScript(path).then(() => {
-    initializer();
+  codeblocks.forEach((e) => {
+    // Large code blocks can cause crashes or slowdowns
+    if (e.innerHTML.length > 30000) {
+      return;
+    }
 
-    codeblocks.forEach((e) => {
-      // Large code blocks can cause crashes or slowdowns
-      if (e.innerHTML.length > 30000) {
-        return;
+    let lang;
+    for (const className of e.classList) {
+      const m = className.match(/^lang-(.+)$/);
+      if (m) {
+        lang = m[1];
+        break;
       }
+    }
+
+    const canHighlight = lang && (lang === "auto" || hljs.getLanguage(lang));
 
-      e.classList.remove("lang-auto");
+    if (canHighlight) {
+      e.classList.remove("lang-auto"); // This isn't a real hljs language. HLJS will warn if it's present, so we remove it.
       hljs.highlightElement(e);
-    });
+    } else {
+      // To make debugging easier, add a data attribute to indicate we skipped it
+      e.dataset.unknownHljsLang = lang;
+    }
   });
 }
 
+async function ensureHighlightJs(langFile) {
+  try {
+    if (!hljsLoadPromise) {
+      hljsLoadPromise = loadHighlightJs(langFile);
+      waitForPromise(hljsLoadPromise);
+    }
+    return await hljsLoadPromise;
+  } catch (e) {
+    // Allow retry next time
+    hljsLoadPromise = null;
+    throw e;
+  }
+}
+
+async function loadHighlightJs(langFile) {
+  const [hljsModule, languageInitializer] = await Promise.all([
+    import("highlight.js/lib/core"),
+    loadLanguageInitializer(langFile),
+  ]);
+
+  const hljs = hljsModule.default;
+
+  languageInitializer(hljs);
+
+  initializer(hljs);
+
+  return hljs;
+}
```

Comment:
This is the key improvement here - we're using webpack-managed `import()` to load hljs core